sqld/src/replication/replica/injector/mod.rs

```diff
@@ -0,0 +1,220 @@
+use std::collections::VecDeque;
+use std::path::Path;
+use std::sync::Arc;
+
+use parking_lot::Mutex;
+use rusqlite::OpenFlags;
+
+use crate::replication::frame::Frame;
+use crate::{replication::FrameNo, DEFAULT_AUTO_CHECKPOINT};
+
+use hook::{
+    InjectorHookCtx, INJECTOR_METHODS, LIBSQL_INJECT_FATAL, LIBSQL_INJECT_OK, LIBSQL_INJECT_OK_TXN,
+};
+
+mod headers;
+mod hook;
+
+#[derive(Debug)]
+pub enum InjectError {}
+
+pub type FrameBuffer = Arc<Mutex<VecDeque<Frame>>>;
+
+pub struct Injector {
+    /// The injector is in a transaction state
+    is_txn: bool,
+    /// Buffer for holding current transaction frames
+    buffer: FrameBuffer,
+    /// Maximum capacity of the frame buffer
+    capacity: usize,
+    /// Injector connection
+    // connection must be dropped before the hook context
+    connection: Arc<Mutex<sqld_libsql_bindings::Connection<'static>>>,
+    /// Pointer to the hook
+    _hook_ctx: Box<InjectorHookCtx>,
+}
+
+/// Methods from this trait are called before and after performing a frame injection.
+/// This trait trait is used to record the last committed frame_no to the log.
+/// The implementer can persist the pre and post commit frame no, and compare them in the event of
+/// a crash; if the pre and post commit frame_no don't match, then the log may be corrupted.
+impl Injector {
+    pub fn new(path: &Path, buffer_capacity: usize) -> crate::Result<Self> {
+        let buffer = FrameBuffer::default();
+        let ctx = InjectorHookCtx::new(buffer.clone());
+        let mut ctx = Box::new(ctx);
+        std::fs::create_dir_all(&path)?;
+        let connection = sqld_libsql_bindings::Connection::open(
+            path,
+            OpenFlags::SQLITE_OPEN_READ_WRITE
+                | OpenFlags::SQLITE_OPEN_CREATE
+                | OpenFlags::SQLITE_OPEN_URI
+                | OpenFlags::SQLITE_OPEN_NO_MUTEX,
```

Comment:
note: once we migrate to `libsql` crate, we need to audit that NO_MUTEX is still ok, because there's been discussions around which threading model we should use.